diff --git a/src/httpd.c b/src/httpd.c
index edaa980c..8ccc515f 100644
--- a/src/httpd.c
+++ b/src/httpd.c
@@ -1448,6 +1448,8 @@ httpd_send_reply(struct evhttp_request *req, int code, const char *reason, struc
     {
       evhttp_send_reply(req, code, reason, evbuf);
     }
+
+   __gcov_flush();
 }
 
 // This is a modified version of evhttp_send_error (credit libevent)
diff --git a/src/main.c b/src/main.c
index d83e45b4..92ad6dfe 100644
--- a/src/main.c
+++ b/src/main.c
@@ -84,6 +84,14 @@ struct event_base *evbase_main;
 static struct event *sig_event;
 static int main_exit;
 
+extern void __gcov_flush();
+
+static void catch_function(int signo)
+{
+  __gcov_flush();
+  _exit(0);
+}
+
 static void
 version(void)
 {
@@ -536,6 +544,12 @@ main(int argc, char **argv)
   mdns_no_cname = false;
   mdns_no_web = false;
 
+  signal(SIGUSR1, catch_function);
+  signal(SIGINT, catch_function);
+  signal(SIGSEGV, catch_function);
+  signal(SIGABRT, catch_function);
+  signal(SIGTERM, catch_function);
+
   while ((option = getopt_long(argc, argv, "D:d:c:P:fb:vw:", option_map, NULL)) != -1)
     {
       switch (option)
